librdmacm/cma.c

```diff
@@ -514,10 +515,11 @@ struct ibv_context **rdma_get_devices(int *num_devices)
 
 		/* reinit newly added devices */
 		if (ucma_init_device(dev)) {
-			/* Couldn't initialize the device: mark it dead and continue */
-			dev->is_device_dead = true;
+			/* Couldn't initialize the device: mark it unavailable and continue. */
+			dev->is_device_fail_to_init = true;
```

Comment:
This change gives same logic as was before, just with another variable name.

Reply:
I think is_device_dead means the verbs is invalid and should be destroyed and no need to retry init further more.

is_device_fail_to_init means we should try reinit the verbs so the user can see this verbs after the init completed.

Reply:
Another approach is to reinit the verbs even if it's already marked dead. Though I think it will make the `is_device_dead` flag ambiguous

Reply:
> I think is_device_dead means the verbs is invalid and should be destroyed and no need to retry init further more.
> 
> is_device_fail_to_init means we should try reinit the verbs so the user can see this verbs after the init completed.

It is a race and library shouldn't try to avoid them. User triggered device removal/addition and later started application, he should synchronize it.

Reply:
There are many races like this and rdma-cm device initialization is maybe most visible.

Reply:
During our test, this is the only barrier to do RDMA device hot plugin. We have to try call ibv_open_device to test if the verbs is valid before call rdma_get_devices.

Reply:
> Without this patch user cannot support hot-plug or hot-unplug without restarting the application.

Are you sure about it? Users who are interested (and using it) in hotplug/unplug listen to udev events to know when device is added and action accordingly.

Reply:
Applications rely on RDMA_CM_EVENT_DEVICE_REMOVAL event to do hot-unplug.
To do hot-plugging, User can listen to the udev events, or polling rdma devices changes by calling rdma_get_devices or ibv_get_device_list.
No matter in which method, if rdma_get_devices is called immediately after the device appears, there is a chance the according verbs would be marked dead and user will never see this device.

Reply:
The original implementation assumes that the verbs will be invalid forever if init failed, but it's not true.

Reply:
What about this change?

diff --git a/librdmacm/cma.c b/librdmacm/cma.c
index 7b924bd0d..0a0da1cce 100644
--- a/librdmacm/cma.c
+++ b/librdmacm/cma.c
@@ -485,10 +485,7 @@ static int ucma_init_all(void)
                if (dev->is_device_dead)
                        continue;
 
-               if (ucma_init_device(dev)) {
-                       /* Couldn't initialize the device: mark it dead and continue */
-                       dev->is_device_dead = true;
-               }
+               ucma_init_device(dev);
        }
        pthread_mutex_unlock(&mut);
        return 0;

Reply:
It works too. ucma_init_device checks cma_dev->verbs to avoid duplicated init.
should I update this PR？

Reply:
I think a flag is still required to hide the uninitilized cma_devs.

Reply:
> Oh wait. this cma_dev should not return to user before init done

Good point.

Reply:
> I think a flag is still required to hide the uninitilized cma_devs.

And I am not convinced yet that this is librdmacm issue.

Reply:
Only librdmacm can see the newly created cma_dev and it just leaves them with an uninitialized ib_verbs. If librdmacm does not retry init, no one can use the device.

Reply:
This patch gives a chance to the user to call rdma_get_devices to retry init the cma_dev.

Reply:
This is a patch I submitted to SPDK to recreate IB listeners after device was hotplugged. I have no alternative but to do it with a workaround.
https://review.spdk.io/gerrit/c/spdk/spdk/+/15616

<img width="780" alt="image" src="https://user-images.githubusercontent.com/5665404/224673702-731a7ff1-8c48-4f10-bfde-1382925c980f.png">